Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in PAL

cdumez opened the pull request:
#### 3798bfb9a809f6d40f35c6ea29226183c07c17d8
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in PAL
<a href="https://bugs.webkit.org/show_bug.cgi?id=282370">https://bugs.webkit.org/show_bug.cgi?id=282370</a>

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/text/icu/UnicodeExtras.h: Added.
* Source/WebCore/PAL/pal/text/DecodeEscapeSequences.h:
(PAL::URLEscapeSequence::decodeRun):
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
* Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
* Source/WebCore/PAL/pal/text/TextEncoding.cpp:
(PAL::UTF7Encoding):
(PAL::TextEncoding::TextEncoding):
(PAL::TextEncoding::domName const):
(PAL::TextEncoding::usesVisualOrdering const):
(PAL::ASCIIEncoding):
(PAL::Latin1Encoding):
(PAL::UTF16BigEndianEncoding):
(PAL::UTF16LittleEndianEncoding):
(PAL::UTF8Encoding):
(PAL::WindowsLatin1Encoding):
* Source/WebCore/PAL/pal/text/TextEncoding.h:
* Source/WebCore/PAL/pal/text/TextEncodingDetectorICU.cpp:
(PAL::detectTextEncoding):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::TextEncodingNameHash::equal):
(PAL::TextEncodingNameHash::hash):
(PAL::HashTranslatorTextEncodingName::hash):
(PAL::HashTranslatorTextEncodingName::equal):
(PAL::atomCanonicalTextEncodingName):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.h:

Canonical link: <a href="https://commits.webkit.org/285998@main">https://commits.webkit.org/285998@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cedc392bae4363c3feab5ebf4566e0fcac93703e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74447 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53876 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78831 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25685 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76564 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1661 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58498 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16816 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48659 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64019 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38904 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45672 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21515 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24018 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67584 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67062 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21861 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80348 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73705 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1024 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66784 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1912 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64037 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66067 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16412 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10004 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8164 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95486 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1728 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4516 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20966 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1757 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1745 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1764 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/icu/UnicodeExtras.h:
```diff
@@ -0,0 +1,43 @@
+
+namespace WTF {
+
+inline std::span<const UCharsetMatch*> ucsdet_detectAll_span(UCharsetDetector* detector, UErrorCode* status)
```
Why inline? Just to save on exports?

cdumez replied:
Do you mean vs moving the implementation to a new cpp file? Because at the moment it is defined in the header so I think I need `inline` for it to build.

darinadler replied:
Yes

cdumez replied:
Ok, I moved it out of line.

cdumez replied:
My main concern was to not risk hurting performance by adding a safe wrapper. In some cases, inlining such "safe" wrappers makes it more likely the compiler will be able to optimize out bounds checks. I don't think it's the case here though.

darinadler commented on Source/WTF/wtf/text/icu/UnicodeExtras.h:
```diff
@@ -0,0 +1,43 @@
+{
+    int32_t matchesCount = 0;
+    auto* matches = ucsdet_detectAll(detector, &matchesCount, status);
+    return unsafeForgeSpan(matches, std::max<int32_t>(matchesCount, 0));
```
The `<int32_t>` here should not be necessary.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
```diff
@@ -31,11 +31,9 @@
 namespace PAL {
 
-static const UChar latin1ConversionTable[256] = {
+static const std::array<UChar, 256> latin1ConversionTable = {
```
Maybe `constexpr` instead of `const`, too?

darinadler commented on Source/WebCore/PAL/pal/text/TextEncoding.h:
```diff
@@ -37,6 +37,7 @@ class TextEncoding : public WTF::URLTextEncoding {
 public:
     TextEncoding() = default;
     PAL_EXPORT TextEncoding(const char* name);
```
Do we still need this one?

cdumez replied:
I will double check but as I remember it, there were still call sites that required it.

cdumez replied:
Actually, I can remove this with minor amount of work.

darinadler commented on Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
```diff
@@ -48,59 +48,64 @@
 // Hash for all-ASCII strings that does case folding.
 struct TextEncodingNameHash {
-    static bool equal(const char* s1, const char* s2)
+    static bool equal(std::span<const LChar> s1, std::span<const LChar> s2)
```
Is `LChar` the best type for characters in all-ASCII strings? I guess maybe.

darinadler commented on Source/WebCore/PAL/pal/text/TextEncodingRegistry.h:
```diff
@@ -43,6 +43,7 @@ PAL_EXPORT std::unique_ptr<TextCodec> newTextCodec(const TextEncoding&);
 
 // Only TextEncoding should use the following functions directly.
 ASCIILiteral atomCanonicalTextEncodingName(const char* alias);
```
Do we still need this one?

cdumez replied:
I think so because I was hoping to remove it and couldn't easily. But I'll double check.

cdumez replied:
Yes, if I drop the constructor above, I can drop this one too.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in PAL
https://bugs.webkit.org/show_bug.cgi?id=282370

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/text/icu/UnicodeExtras.h: Added.
* Source/WebCore/PAL/pal/text/DecodeEscapeSequences.h:
(PAL::URLEscapeSequence::decodeRun):
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
* Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
* Source/WebCore/PAL/pal/text/TextEncoding.cpp:
(PAL::UTF7Encoding):
(PAL::TextEncoding::TextEncoding):
(PAL::TextEncoding::domName const):
(PAL::TextEncoding::usesVisualOrdering const):
(PAL::ASCIIEncoding):
(PAL::Latin1Encoding):
(PAL::UTF16BigEndianEncoding):
(PAL::UTF16LittleEndianEncoding):
(PAL::UTF8Encoding):
(PAL::WindowsLatin1Encoding):
* Source/WebCore/PAL/pal/text/TextEncoding.h:
* Source/WebCore/PAL/pal/text/TextEncodingDetectorICU.cpp:
(PAL::detectTextEncoding):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::TextEncodingNameHash::equal):
(PAL::TextEncodingNameHash::hash):
(PAL::HashTranslatorTextEncodingName::hash):
(PAL::HashTranslatorTextEncodingName::equal):
(PAL::atomCanonicalTextEncodingName):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.h:

Canonical link: https://commits.webkit.org/285998@main
